### PR TITLE
Move mongo field type logic into mongoadapter

### PR DESCRIPTION
**Pull request by drew-gross**

### CODE REVIEW

File: src/Schema.js [hunk 1]
```diff
@@ -591,14 +582,14 @@ class Schema {
...
     if (perms['*']) {
       return Promise.resolve();
     }
     // Check permissions against the aclGroup provided (array of userId/roles)
-    var found = false;
-    for (var i = 0; i < aclGroup.length && !found; i++) {
+    let found = false;
+    for (let i = 0; i < aclGroup.length && !found; i++) {
       if (perms[aclGroup[i]]) {
         found = true;
```
Comment (blacha) [inline src/Schema.js]: do you want to break the loop early if it was found?

File: src/Schema.js [hunk 2]
```diff
@@ -169,47 +169,37 @@ function invalidClassNameMessage(className) {
...
+  'Object',
+  'Array',
+  'GeoPoint',
+  'File',
+];
+// Returns an error suitable for throwing if the type is invalid
+const fieldTypeIsInvalid = ({ type, targetClass }) => {
+  if (['Pointer', 'Relation'].includes(type)) {
+    if (!targetClass) {
+      return new Parse.Error(135, `type ${type} needs a class name`);
```
Comment (blacha) [inline src/Schema.js]: Why is this error code a magic number not a ``Parse.Error.SOMETHING``

Comment (drew-gross) [inline src/Schema.js]: The error codes come from the Parse JS SDK which doesn't include error 135 (invalid type) :(

File: src/Schema.js (hunk 1 above)
Comment (drew-gross) [inline src/Schema.js]: Or even better: use `Array.prototype.some`. But I'd rather do that in a different diff.

File: src/Schema.js [hunk 3]
```diff
@@ -244,9 +234,8 @@ class Schema {
         // createdAt and updatedAt are wacky and have legacy baggage
         parseFormatSchema.createdAt = { type: 'String' };
         parseFormatSchema.updatedAt = { type: 'String' };
-        this.data[schema.className] = _.mapValues(parseFormatSchema, parseField =>
-          schemaAPITypeToMongoFieldType(parseField).result
-        );
+        //Necessary because we still use the mongo type internally here :(
+        this.data[schema.className] = _.mapValues(parseFormatSchema, MongoSchemaCollection._DONOTUSEparseFieldTypeToMongoFieldType);
```
Comment (flovilmart) [inline src/Schema.js]: I'd rather have an internal method that is OK to use internally, and an external one that may be removed later. What do you think?

File: src/Schema.js [hunk 4]
```diff
@@ -397,72 +386,75 @@ class Schema {
...
+        // The update failed. This can be okay - it might have been a race
+        // condition where another client updated the schema in the same
+        // way that we wanted to. So, just reload the schema
+        return this.reloadData();
+      }).then(() => {
+        // Ensure that the schema now validates
+        return this.validateField(className, fieldName, type, true);
+      }, (error) => {
+        // The schema still doesn't validate. Give up
+        console.log(error)
```
Comment (flovilmart) [inline src/Schema.js]: use the logger

File: src/Schema.js (hunk 3 above)
Comment (flovilmart) [inline src/Schema.js]: As _colleciton is the MongoSchemaCollection, I propose that you do that transform in getAllSchemas() ? Later PR?

**Comment by flovilmart** [comment]
LGTM.

File: src/Schema.js (hunk 4 above)
Comment (drew-gross) [inline src/Schema.js]: Will do

File: src/Schema.js (hunk 3 above)
Comment (drew-gross) [inline src/Schema.js]: Yep, later PR

**Pull request merged**